graphql-ballerina/duplicate_field_remover.bal

```diff
@@ -0,0 +1,181 @@
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import graphql.parser;
+
+class DuplicateFieldRemover {
+    private parser:DocumentNode documentNode;
+
+    public isolated function init(parser:DocumentNode documentNode) {
+        self.documentNode = documentNode;
+    }
+
+    public isolated function remove() {
+        self.visitDocument(self.documentNode);
+    }
+
+    public isolated function visitDocument(parser:DocumentNode documentNode) {
+        parser:OperationNode[] operations = documentNode.getOperations();
+        foreach parser:OperationNode operationNode in operations {
+            self.visitOperation(operationNode);
+        }
+    }
+
+    public isolated function visitOperation(parser:OperationNode operationNode) {
+        self.visitSelection(operationNode.getSelections());
+        self.visitField(operationNode.getFields());
+        self.visitFragment(operationNode.getFragments());
+    }
+
+    public isolated function visitSelection(parser:Selection[] selections) {
+        self.removeDuplicateSelections(selections);
+        foreach parser:Selection selection in selections {
+            if (selection.isFragment) {
+                parser:FragmentNode subFragmentNode = self.documentNode.getFragments().get(selection.name);
+                self.visitSelection(subFragmentNode.getSelections());
+            } else {
+                parser:FieldNode subFieldNode = <parser:FieldNode>selection?.node;
+                self.visitSelection(subFieldNode.getSelections());
+            }
+        }
+    }
+
+    public isolated function visitField(parser:FieldNode[] fields) {
+        self.removeDuplicateFields(fields);
+        foreach parser:FieldNode fieldNode in fields {
+            self.visitField(fieldNode.getFields());
+            self.visitFragment(fieldNode.getFragments());
+        }
+    }
+
+    public isolated function visitFragment(string[] fragments) {
+        self.removeDuplicateFragments(fragments);
+        foreach string fragment in fragments {
+            parser:FragmentNode fragmentNode = self.documentNode.getFragments().get(fragment);
+            self.visitFragment(fragmentNode.getFragments());
+            self.visitField(fragmentNode.getFields());
+        }
+    }
+
+    private isolated function removeDuplicateSelections(parser:Selection[] selections) {
+        map<parser:Selection> visitedSelections = {};
+        map<parser:Selection> visitedFragmentOnTypes = {};
+        int[] duplicateIndexes = [];
+        int i = 0;
+        while (i < selections.length()) {
+            parser:Selection subSelection = selections[i];
```

Comment:
Usually the prefix `sub` is used when there is a parent of the same type that exists within the scope. Otherwise, there's no need to use `sub` here.

Check the other places as well.

```suggestion
            parser:Selection selection = selections[i];
```